miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/storage/DownloadedManifestCache.kt

```diff
@@ -0,0 +1,108 @@
+package com.rakuten.tech.mobile.miniapp.storage
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.util.Log
+import androidx.annotation.VisibleForTesting
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import com.rakuten.tech.mobile.miniapp.permission.MiniAppCustomPermission
+import com.rakuten.tech.mobile.miniapp.permission.MiniAppCustomPermissionResult
+import com.rakuten.tech.mobile.miniapp.permission.MiniAppCustomPermissionType
+import kotlin.Exception
+
+/**
+ * A caching class to read and store the [CachedManifest] per MiniApp using [SharedPreferences].
+ */
+@Suppress("TooGenericExceptionCaught", "SwallowedException")
+internal class DownloadedManifestCache(
+    context: Context
+) {
+    private val prefs: SharedPreferences = context.getSharedPreferences(
+        "com.rakuten.tech.mobile.miniapp.manifest.cache.downloaded", Context.MODE_PRIVATE
+    )
+
+    /**
+     * Reads the downloaded manifest from SharedPreferences.
+     * @param [miniAppId] the first half of key provided to find the stored manifest per MiniApp.
+     * @param [versionId] the second half of key provided to find the stored manifest per MiniApp.
+     * @return [CachedManifest] an object to contain MiniAppManifest per versionId,
+     * if data has been stored in cache, otherwise null.
+     */
+    fun readDownloadedManifest(miniAppId: String, versionId: String): CachedManifest? {
+        val manifestJsonStr =
+            prefs.getString(primaryKey(miniAppId, versionId), null) ?: return null
+        return try {
+            Gson().fromJson(manifestJsonStr, object : TypeToken<CachedManifest>() {}.type)
+        } catch (e: Exception) {
+            Log.e(this::class.java.canonicalName, e.message.toString())
+            null
+        }
+    }
+
+    fun storeDownloadedManifest(
+        miniAppId: String,
+        cachedManifest: CachedManifest
+    ) {
+        val jsonToStore: String = Gson().toJson(cachedManifest)
+        prefs.edit().putString(primaryKey(miniAppId, cachedManifest.versionId), jsonToStore).apply()
```

Comment:
We already have the `versionId` in the `CachedManifest` object, so why do you need set the version as part of the key?

I think we should store the manifest using only the `miniAppId` as the key. We only want to store one manifest for each mini app id, and it should be the manifest for the currently downloaded version. Also, this is causing problems because it means you need to pass the `versionId` every time you retrieve the manifest, but the `versionId` isn't relevant when retrieving the manifest.

Reply:
Right, using `versionId` in this case causes issues. I have updated this part.